devito/grid.py

```diff
@@ -209,3 +258,118 @@ def __setstate__(self, state):
+                                                              thickness_left,
+                                                              thickness_right))
+                except ValueError:
+                    side, thickness = v
+                    if side == 'left':
+                        sub_dimensions.append(SubDimension.left('%sleft' % k.name, k,
+                                                                thickness))
+                    elif side == 'right':
+                        sub_dimensions.append(SubDimension.right('%sright' % k.name, k,
+                                                                 thickness))
+                    else:
+                        raise ValueError("Expected sides 'left|right', not `%s`" % side)
+        self._dimensions = tuple(sub_dimensions)
+
+        # Compute the SubDomain shape
+        self._shape = tuple(s - (sum(d.thickness_map.values()) if d.is_Sub else 0)
+                            for d, s in zip(self._dimensions, shape))
+
+    def __eq__(self, other):
+        if not isinstance(other, SubDomain):
+            return False
+        return self.name == other.name and self.dimensions == other.dimensions
+
+    def __hash__(self):
+        return hash((self.name, self.dimensions))
+
+    def __str__(self):
+        return "SubDomain[%s]" % self.name
+
+    __repr__ = __str__
+
+    @property
+    def finalized(self):
+        return self._dimensions is not None
+
+    @property
+    def dimensions(self):
+        return self._dimensions
+
+    @property
+    def dimension_map(self):
+        return {d.parent: d for d in self.dimensions}
+
+    @property
+    def shape(self):
+        return self._shape
+
+    def define(self, dimensions):
+        """
+        Return a dictionary ``M : D -> V``, where: ::
+
+            * D are the Grid dimensions
+            * M(d) = {d, ('left', int), ('middle', int, int), ('right', int, int)}.
+              If ``M(d) = d``, the SubDomain spans the entire domain along the
+              :class:`Dimension` ``d``. In all other cases, the SubDomain spans
+              a contiguous subregion of the domain. For example, if
+              ``M(d) = ('left', 4)``, The SubDomain has thickness 4 near ``d``'s
+              left extreme.
+
+        .. note::
+
+            This method should be overridden by each subclass of SubDomain that
+            wants to define a new type of subdomain.
+        """
+        raise NotImplementedError
+
+
+class Domain(SubDomain):
+
+    """
+    The entire computational domain (== boundary + interior).
+    """
+
+    name = 'domain'
+
+    def define(self, dimensions):
+        return dict(zip(dimensions, dimensions))
+
+
+class Interior(SubDomain):
+
+    """
+    The interior of the computational domain (i.e., boundaries are excluded).
+    """
+
+    name = 'interior'
+
+    def define(self, dimensions):
+        return {d: ('middle', 1, 1) for d in dimensions}
```

Comment:
what's these hard coded `1,1`

Reply:
this is for the Interior subdomain, which is one of the preexisting subdomains shipped by Devito. And the interior subdomain cuts away the extremes, so 1 on the left and 1 on the right side

Reply:
Why isn't default `0`. Default case shouldn't be the domain is the Grid without any halo?

Reply:
that's for interior, and since ``domain = boundary + interior``, with the two 1s we are stripping off the boundary at left and right